modules/pallets/messaging-fees/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+		type PriceOracle: PriceOracle<Self::Balance>;
+
+		/// The target message size
+		#[pallet::constant]
+		type TargetMessageSize: Get<u32>;
+
+		/// Weight information for operations
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Total bytes processed in the current epoch for all chains
+	#[pallet::storage]
+	#[pallet::getter(fn total_bytes_processed)]
+	pub type TotalBytesProcessed<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	/// Stores whitelisted routes for incentives. The key is a tuple of (source, destination).
+	#[pallet::storage]
+	#[pallet::getter(fn incentivized_routes)]
+	pub type IncentivizedRoutes<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		StateMachine,
+		Twox64Concat,
+		StateMachine,
+		bool,
+		OptionQuery,
+	>;
+
+	/// Current active Epoch for incentivization
+	#[pallet::storage]
+	#[pallet::getter(fn epoch)]
+	pub type Epoch<T: Config> = StorageValue<_, EpochInfo<BlockNumberFor<T>>, ValueQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Reward transfer failed
+		RewardTransferFailed,
+		/// Calculation overflow
+		CalculationOverflow,
+		/// Oracle Price conversion error
+		ErrorInPriceConversion,
+		/// State machine per byte fee not found
+		PerByteFeeNotFound,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// State machine Routes supported for incentives
+		RouteSupported { source_chain: StateMachine, destination_chain: StateMachine },
+		/// A relayer was rewarded
+		RelayerRewarded {
+			/// Relayer account that received the reward
+			relayer: T::AccountId,
+			/// Amount of the reward
+			amount: <T as pallet_ismp::Config>::Balance,
+		},
+		/// A relayer was charged a fee
+		RelayerCharged {
+			/// Relayer account that was charged
+			relayer: T::AccountId,
+			/// Amount of the fee
+			amount: <T as pallet_ismp::Config>::Balance,
+		},
+		/// A new epoch has started
+		NewEpoch {
+			/// The index of the new epoch
+			index: u64,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Whitelists a route for messaging fees.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::set_supported_route())]
+		pub fn set_supported_route(
+			origin: OriginFor<T>,
+			source: StateMachine,
+			destination: StateMachine,
+		) -> DispatchResult {
+			<T as Config>::IncentivesOrigin::ensure_origin(origin)?;
+
+			IncentivizedRoutes::<T>::insert(source, destination, true);
+
+			Self::deposit_event(Event::<T>::RouteSupported {
+				source_chain: source,
+				destination_chain: destination,
+			});
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> pallet_session::SessionHandler<T::AccountId> for Pallet<T> {
+	const KEY_TYPE_IDS: &'static [KeyTypeId] = &[];
+
+	fn on_genesis_session<Ks: OpaqueKeys>(_validators: &[(T::AccountId, Ks)]) {}
+
+	fn on_new_session<Ks: OpaqueKeys>(
+		_changed: bool,
+		_validators: &[(T::AccountId, Ks)],
+		_queued_validators: &[(T::AccountId, Ks)],
+	) {
```

Comment:
If `changed != true` early return. [From the docs](https://docs.rs/pallet-session/latest/pallet_session/trait.SessionHandler.html#tymethod.on_new_session)